# MiniAtlas <img src="./images/MiniAtlas_logo.jpg" align="right" width="200"/>
MiniAtlas is a single-nuclei, multi-omic atlas profiling transcriptome + histone modification in adult human brain.
<br/><br/>

ðŸ¥ƒ This repository contains code to reproduce the analysis for our manuscript: **Single-Cell Atlas of Transcription and Chromatin States Reveals Regulatory Programs in the Human Brain**
<br/><br/>

## Abstract
Understanding how cell-type-specific regulatory programs are established and maintained in the human brain requires direct measurement of chromatin states alongside transcription. Here, we present a large-scale single-cell multimodal atlas of the adult human brain generated by jointly profiling transcriptomes with active (H3K27ac) and repressive (H3K27me3) histone modifications. Across 18 brain regions, we profile over 750,000 nuclei spanning 160 cell types and integrate these data with single-cell chromatin accessibility, DNA methylation, genome architecture, and spatial transcriptomic datasets. This integrated framework enables functional annotation of over 500,000 candidate cis-regulatory elements, distinguishing accessible, active, and Polycomb-repressed states with cell-type- and regional specificity. We further link enhancers to target genes, infer active or repressive gene regulatory networks in a cell-type-resolved manner, and functionally annotate chromatin loops, revealing neuron-enriched super long-range repressive interactions that preferentially engage developmental genes. By coupling these regulatory maps with genetic association data and sequence-based modeling, we prioritize noncoding variants, nominate their effector genes and vulnerable cell types across neuropsychiatric disorders. Cross-species integration further reveals an evolutionary asymmetry in regulatory grammar, with active programs broadly conserved and repressive programs more divergent. Together, this resource establishes a functional reference for the human brain regulatory architecture, providing a framework for interpreting noncoding genetic variants, understanding epigenetic memory, and linking chromatin states to brain structure and disease.
<br/><br/>

![graphic_abstract](./images/Graphic_abstract.jpeg)

## Method


## Analysis
### [01.pre_process](https://github.com/Xieeeee/FNIH_Heart/tree/main/01.pre-process)
This directory contains code for pre-processing data from different modalities. 
### [02.analysis](https://github.com/Xieeeee/FNIH_Heart/tree/main/02.analysis)
This directory contains code for re-producing figures in our manuscript. 

## Data
All processed and raw data will be available at NCBI Gene Expression Omnibus (GEO) and NeMO data archive once the manuscript is published
